### ISSUE TITLE
Bad Performance with Joins Due to n+1 calls of Filter in Virtual Tables

### ISSUE THREAD

**Issue by cancan101** [rogerbinns/apsw#331]
Reposting from https://github.com/betodealmeida/shillelagh/issues/224 as the issue / usage question might better belong to apsw

I am trying to join two virtual tables with a sql query like:
```sql
    select 
        o.[abc]
        from 
            O o join C c
        on
            c.[id] = o.[c_id]
```
however that results in `n` calls to C's `xFilter` where `n` is the number of rows in `O`. I would think that sqlite would keep the data in memory itself rather than repeatedly calling `Filter`. And if not by default to avoid allocating extra memort, is there a way to tell sqlite to "cache" the initial call?

Some potentially relevant posts:
- https://sqlite-users.sqlite.narkive.com/cVmIro4J/sqlite-question-on-join-in-virtual-tables
- https://stackoverflow.com/questions/60571296/speeding-up-a-sqlite-query-with-a-join-to-a-virtual-table

Using a materialized CTE also did _not_ seem to fix the issue:
```sql
with CC as materialized (select * from C)
```

**Comment by cancan101** [comment]
It might not be an issue with apsw or even solvable through APSW. That being said, I wanted to see if there was some way through the sqlite interface to tell sqlite to keep its own representation of full table rather than it having to "re-pull" it from the Virtual Table.

Basically the issue is that each time sqlite decides to run a full table (sequential) scan on a virtual table, my `VTCursor.Filter` for that `VTable` is getting called. In the case of a join between tables `O` and `C`, that means `n+1` calls of `Filter` (where `n` is rows in the table the query planner decides to iterate over). This seems in line with what sqlite would do with a non virtual table (scanning the full table), but now with the downside of that data coming from a Python function.

**Comment by cancan101** [comment]
ok, makes sense that this is outside the scope of apsw. I have posted the issue upstream with sqlite: https://sqlite.org/forum/forumpost/7e2802db01

I have used `EXPLAIN QUERY PLAN` and what it intends to do makes sense, which is two full table scans. What doesn't make sense is that it isn't "materialized" the tables during the lifetime of the query. Essentially I do want the materialized table as you pointed out but I just want it to last for the query. I tried to get this using `with CC as materialized (select * from C)` but that didn't work. Likewise with a "temp table". I would want them to last just for the query lifetime.
